mllib/src/main/scala/org/apache/spark/mllib/clustering/metrics/FastEuclideanOps.scala

```diff
@@ -0,0 +1,77 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.mllib.clustering.metrics
+
+import breeze.linalg.{ DenseVector => BDV, SparseVector => BSV, Vector => BV }
+
+import org.apache.spark.mllib.base._
+import org.apache.spark.mllib.linalg.{ SparseVector, DenseVector, Vector }
+import org.apache.spark.mllib.base.{ Centroid, FPoint, PointOps, Infinity, Zero }
+
+class FastEUPoint(raw: BV[Double], weight: Double) extends FPoint(raw, weight) {
+  val norm = if (weight == Zero) Zero else raw.dot(raw) / (weight * weight)
```

Comment:
Simply put, computing the distance (and the cached values per point and centered used in the distance) is faster in homogeneous coordinates. 

Imagine computing dot(x,y), where x and y are vectors. If x and y are in homogeneous coordinates, then it is faster to perform the fit product in homogeneous coordinates then divide by the weights of the two vectors than it is to divide each homogenous coordinate by the weights first and the do the dot product.

Sent from my iPhone

> On Jan 19, 2015, at 5:02 PM, Xiangrui Meng <notifications@github.com> wrote:
> 
> In mllib/src/main/scala/org/apache/spark/mllib/clustering/metrics/FastEuclideanOps.scala:
> 
> > + * distributed under the License is distributed on an "AS IS" BASIS,
> > + * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
> > + * See the License for the specific language governing permissions and
> > + * limitations under the License.
> > + */
> > +
> > +package org.apache.spark.mllib.clustering.metrics
> > +
> > +import breeze.linalg.{ DenseVector => BDV, SparseVector => BSV, Vector => BV }
> > +
> > +import org.apache.spark.mllib.base._
> > +import org.apache.spark.mllib.linalg.{ SparseVector, DenseVector, Vector }
> > +import org.apache.spark.mllib.base.{ Centroid, FPoint, PointOps, Infinity, Zero }
> > +
> > +class FastEUPoint(raw: BV[Double], weight: Double) extends FPoint(raw, weight) {
> > +  val norm = if (weight == Zero) Zero else raw.dot(raw) / (weight * weight)
> Could you elaborate more about the benefit of using homogeneous coordinates? It should be at least documented that what "weight" stands for here. If a cluster have 100 points, do you assign its center with weight 100? If so, how to compute the distance between this center and an input point?
> 
> —
> Reply to this email directly or view it on GitHub.
>